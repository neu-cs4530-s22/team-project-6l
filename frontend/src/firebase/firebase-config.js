// Import the functions you need from the SDKs you need
import { initializeApp } from "firebase/app";
import { getAuth } from "firebase/auth";
// TODO: Add SDKs for Firebase products that you want to use
// https://firebase.google.com/docs/web/setup#available-libraries

// Your web app's Firebase configuration
// For Firebase JS SDK v7.20.0 and later, measurementId is optional
const firebaseConfig = {
  apiKey: "AIzaSyCY6FF-e9h48NmhLu3WrHdWOzn7bjN_i1g",
  authDomain: "coveytown-ac1e3.firebaseapp.com",
  projectId: "coveytown-ac1e3",
  storageBucket: "coveytown-ac1e3.appspot.com",
  messagingSenderId: "987672035225",
  appId: "1:987672035225:web:094db237029f4c824fdd08",
  measurementId: "G-6E85LP2QW8"
};

// Initialize Firebase
const app = initializeApp(firebaseConfig);
const auth = getAuth(app);
export default auth;