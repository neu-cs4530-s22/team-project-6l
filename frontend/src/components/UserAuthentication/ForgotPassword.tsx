import React from 'react';
import {
  Box,
  FormControl,
  FormLabel,
  Input,
  Button,
  Text,
  Flex,
  Alert,
  AlertIcon,
  AlertTitle,
  CloseButton,
} from '@chakra-ui/react';
import { useHistory } from 'react-router-dom';
import { sendPasswordResetEmail } from 'firebase/auth';
import auth from '../../firebaseAuth/firebase-config';
import authCheck from './authCheck';

/**
 * Display forgot password form, 
 * which will send a reset link to email if the users forget their password
 * and return error message if email is yet registered or invalid.
 */
export default function ForgotPassword():JSX.Element {
  const history = useHistory();
  const [email, setEmail] = React.useState('');
  const [isSent, setSent] = React.useState(false);
  const [isAlert, setAlert] = React.useState(false);
  const [alertMess, setAlertMess] = React.useState('');

  const onSignInClick = ((event: React.MouseEvent) => {
    event.preventDefault();
    history.push("/");
  });

  const onSendingClick = async (event: React.MouseEvent) => {
    event.preventDefault();
    await sendPasswordResetEmail(auth, email.trim())
    .then(() => setSent(true))
    .catch(error => {
      const {code} = error;
      setAlert(true);
      setAlertMess(authCheck(code));
    })
  }

  return (
    <Flex height="100vh" width="full" justifyContent="center" align="center" flexDirection="column">
      <Box
        textAlign="left"
        w="800px" maxW='lg'
        backgroundColor="white"
        borderRadius='lg'
        paddingTop="6"
        paddingBottom="6"
        paddingLeft="8"
        paddingRight="8"
        shadow="0 0 0 2px rgba(0, 0, 0, 0.2)"
      >
        {isSent ?
          <>
            <Text data-testid="sent-message" fontSize='sm' fontWeight="semibold">A reset link has been sent to your email.</Text>
            <Button data-testid="signin-btn" mt={4} type="submit" backgroundColor="blue.500" color="white" onClick={e => onSignInClick(e)}>Sign in</Button>
          </> :
          <>
            <Text fontSize='2xl' fontWeight="semibold" marginTop="2">Forgot Your Password?</Text>
            <Box textAlign="left" marginTop="2">
              <FormControl>
                <FormLabel>Enter your email</FormLabel>
                <Input data-testid="forgot-email" type="email" placeholder='Email' onChange={(event) => setEmail(event.target.value)} />
              </FormControl>
            </Box>
            {isAlert ?
            <Box marginTop="2">
              <Alert data-testid="error" status='error'>
                <AlertIcon />
                <AlertTitle data-testid="error-message" mr={2}>{alertMess}</AlertTitle>
                <CloseButton marginLeft="1" position='absolute' right='8px' top='8px' onClick={() => setAlert(false)} />
              </Alert>
            </Box> : <></>}
            <Box justifyContent="flex-end" display="flex">
              <Button mt={4} type="submit" backgroundColor="blue.500" color="white" onClick={e => onSendingClick(e)}> Reset </Button>
            </Box>
          </>
        }
      </Box>
    </Flex>
  );
}