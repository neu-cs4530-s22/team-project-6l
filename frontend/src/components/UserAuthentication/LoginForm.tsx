import {
  Alert,
  AlertIcon,
  AlertTitle,
  Box,
  Button,
  CloseButton,
  FormControl,
  FormLabel,
  Input,
} from '@chakra-ui/react';
import { FirebaseError } from 'firebase/app';
import { signInWithEmailAndPassword } from 'firebase/auth';
import React from 'react';
import { useHistory } from 'react-router-dom';
import auth from '../../firebaseAuth/firebase-config';
import authCheck from './authCheck';

/**
 * Displays a form which handles user's sign in
 * navigate to prejoinscreen if user succesfully sign up
 * otherwise it will appear alert error which indicates what user can't log in
 */
export default function LoginForm(): JSX.Element {
  const history = useHistory();
  const [email, setEmail] = React.useState('');
  const [password, setPassword] = React.useState('');
  const [isAlert, setAlert] = React.useState(false);
  const [alertMess, setAlertMess] = React.useState('');

  const logInWithEmailAndPassword = async () => {
    try {
      const credentials = await signInWithEmailAndPassword(auth, email, password);
      if (credentials.user) {
        history.push('/pre-join-screen');
      }
    } catch (err) {
      const error = err as FirebaseError;
      const { code } = error;
      setAlert(true);
      setAlertMess(authCheck(code));
    }
  };

  const onLoginClick = (event: React.MouseEvent) => {
    event.preventDefault();
    if (!email || !password) {
      setAlert(true);
      if (!email) setAlertMess('Enter your email');
      else if (!password) setAlertMess('Enter your password');
      return;
    }
    logInWithEmailAndPassword();
  };

  const onForgotPassClick = (event: React.MouseEvent) => {
    event.preventDefault();
    history.push('/forgot-password');
  };

  return (
    <Box textAlign='left' w='400px' marginBottom='2'>
      <FormControl>
        <FormLabel>Email</FormLabel>
        <Input
          data-testid='login-email'
          type='email'
          placeholder='Email'
          onChange={event => setEmail(event.target.value)}
        />
      </FormControl>
      <FormControl>
        <FormLabel>Password</FormLabel>
        <Input
          data-testid='login-password'
          type='password'
          placeholder='Password'
          onChange={event => setPassword(event.target.value)}
        />
      </FormControl>
      <Button
        data-testid='forgot-password-btn'
        fontSize='sm'
        color='blue.500'
        fontWeight='semibold'
        variant='link'
        onClick={e => onForgotPassClick(e)}>
        Forgot Password?
      </Button>
      {isAlert ? (
        <Box marginTop='2'>
          <Alert data-testid='alert-error' status='error'>
            <AlertIcon />
            <AlertTitle data-testid='alert-message' mr={2}>
              {alertMess}
            </AlertTitle>
            <CloseButton
              marginLeft='1'
              position='absolute'
              right='8px'
              top='8px'
              onClick={() => setAlert(false)}
            />
          </Alert>
        </Box>
      ) : (
        <></>
      )}
      <Button
        data-testid='sign-in-btn'
        width='full'
        mt={4}
        type='submit'
        backgroundColor='blue.500'
        color='white'
        onClick={e => onLoginClick(e)}>
        Sign In
      </Button>
    </Box>
  );
}
