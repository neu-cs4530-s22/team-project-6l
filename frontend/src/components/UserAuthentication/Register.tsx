import React from 'react';
import {
  Flex,
  Box,
  Heading,
  FormControl,
  FormLabel,
  Input,
  Button,
} from '@chakra-ui/react';
import { useHistory } from 'react-router-dom';
import {createUserWithEmailAndPassword} from 'firebase/auth';
import auth from '../../firebase/firebase-config';


export default function Register() {
  const history = useHistory();
  const [email, setEmail] = React.useState('');
  const [password, setPassword] = React.useState('');
  const [confirmPassword, setConfirmPassword] = React.useState('');

  const onLogInClick = ((event:React.MouseEvent) => {
    event.preventDefault();
    history.push("/");
  });

  const onRegisterClick = (event:React.MouseEvent) => {
    if (email === '' || password === '' || confirmPassword ==='' || password !== confirmPassword) { 
      alert('invalid inputs');
    } else {
      event.preventDefault();
      createUserWithEmailAndPassword(auth, email, password)
      .then(() => {
        history.push("/")
      })
      .catch(error => alert(error.message));

    }
  }

  return (
    <Flex width="full" align="center" justifyContent="center">
      <Box>
        <Box textAlign="center">
          <Heading>Register</Heading>
        </Box>
        <Box textAlign="left">
          <FormControl>
            <FormLabel>Email</FormLabel>
            <Input id="reg-email" type="email" onChange={(event) => setEmail(event.target.value)}/>
          </FormControl>

          <FormControl>
            <FormLabel>Password</FormLabel>
            <Input id="reg-password" type="password" onChange={(event) => setPassword(event.target.value)}/>
          </FormControl>

          <FormControl>
            <FormLabel>Confirm Password</FormLabel>
            <Input id="reg-confirmpassword" type="password" onChange={(event) => setConfirmPassword(event.target.value)}/>
          </FormControl>

          <Button width="full" mt={4} type="submit" onClick={e => onRegisterClick(e)}>
            Register
          </Button>
        </Box>
        <Box textAlign="center">
          <div>{'Already have an account? '}</div>
          <div>
            <Button variantColor="teal" variant="link" onClick={onLogInClick}>
              Log in 
            </Button>
          </div>
        </Box>
      </Box>
    </Flex>
  );
}