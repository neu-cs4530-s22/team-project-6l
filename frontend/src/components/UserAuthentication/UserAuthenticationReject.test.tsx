import React from 'react';
import '@testing-library/jest-dom';
import { fireEvent, render, waitFor } from '@testing-library/react';
import { sendPasswordResetEmail, signInWithEmailAndPassword, createUserWithEmailAndPassword } from 'firebase/auth';
import ForgotPassword from './ForgotPassword';
import LoginForm from './LoginForm';
import Register from './Register';

const mockHistoryPush = jest.fn();

jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useHistory: () => ({
    push: mockHistoryPush,
  }),
}));

jest.mock('../../firebaseAuth/firebase-config', () => ({
  auth: jest.fn().mockReturnThis(),
}));

jest.mock('firebase/auth', () => ({
  sendPasswordResetEmail: jest.fn().mockRejectedValue(() => Promise.reject()),
  signInWithEmailAndPassword: jest.fn().mockRejectedValue(() => Promise.reject()),
  createUserWithEmailAndPassword: jest.fn().mockRejectedValue(() => Promise.reject()),
  signInWithPopup: jest.fn().mockRejectedValue(() => Promise.reject()),
}));

describe('ForgotPassword', () => {
  afterEach(() => {
    jest.clearAllMocks();
    mockHistoryPush.mockReset();
  });
  it('should not successfully login', async () => {
    const { getByTestId, queryByTestId, container } = render(<LoginForm />);
    const emailEl = getByTestId('login-email');
    const passEl = getByTestId('login-password');
    fireEvent.change(emailEl, { target: { value: 'testing@gmail.com' } });
    fireEvent.change(passEl, { target: { value: '123456' } });
    fireEvent.click(getByTestId('sign-in-btn'));
    await waitFor(() => {
      expect(signInWithEmailAndPassword).toBeCalled();
      expect(queryByTestId('alert-message')).toBeDefined();
    })
    expect(container.firstChild).toMatchSnapshot();
    expect(mockHistoryPush).not.toBeCalled();
  });
  it('should not successfully create a new user', async () => {
    const { getByTestId, queryByTestId } = render(<Register />);
    fireEvent.change(getByTestId('reg-email'), { target: { value: 'testing@gmail.com' } });
    fireEvent.change(getByTestId('reg-password'), { target: { value: '123456' } });
    fireEvent.change(getByTestId('reg-confirmpassword'), { target: { value: '123456' } });
    fireEvent.click(getByTestId('sign-up'));
    await waitFor(() => {
      expect(createUserWithEmailAndPassword).toBeCalled();
      expect(queryByTestId('error')).toBeDefined();
    })
    expect(mockHistoryPush).not.toBeCalled();
  });
  it('should unsuccessfully send the reset email in ForgotPassword', async () => {
    const { getByTestId, getByText, queryByTestId, queryByText, container } = render(<ForgotPassword />);
    const email = getByTestId('forgot-email');
    fireEvent.change(email, { target: { value: 'testing@gmail.com' } });
    const btn = getByText('Reset');
    fireEvent.click(btn);
    await waitFor(() => {
      expect(queryByTestId('error-message')).toBeDefined();
      expect(queryByText('A reset link has been sent to your email.')).toBeNull();
      expect(sendPasswordResetEmail).toBeCalled();
    })
    // the snapshot should show the error message
    expect(container.firstChild).toMatchSnapshot();
  });
}) 