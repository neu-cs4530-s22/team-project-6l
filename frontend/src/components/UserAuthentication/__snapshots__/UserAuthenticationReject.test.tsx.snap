// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ForgotPassword should not successfully login 1`] = `
<div
  class="css-1t57a7l"
>
  <div
    class="chakra-form-control css-0"
    role="group"
  >
    <label
      class="chakra-form__label css-uvho5g"
      for="field-1"
      id="field-1-label"
    >
      Email
    </label>
    <input
      class="chakra-input css-0"
      data-testid="login-email"
      id="field-1"
      placeholder="Email"
      type="email"
    />
  </div>
  <div
    class="chakra-form-control css-0"
    role="group"
  >
    <label
      class="chakra-form__label css-uvho5g"
      for="field-2"
      id="field-2-label"
    >
      Password
    </label>
    <input
      class="chakra-input css-0"
      data-testid="login-password"
      id="field-2"
      placeholder="Password"
      type="password"
    />
  </div>
  <button
    class="chakra-button css-1qgagya"
    data-testid="forgot-password-btn"
    type="button"
  >
    Forgot Password?
  </button>
  <div
    class="css-g89h0y"
  >
    <div
      class="chakra-alert css-1hwd9xd"
      data-testid="alert-error"
      role="alert"
    >
      <span
        class="chakra-alert__icon css-1c6vscd"
      >
        <svg
          class="chakra-icon css-fi0ww0"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M11.983,0a12.206,12.206,0,0,0-8.51,3.653A11.8,11.8,0,0,0,0,12.207,11.779,11.779,0,0,0,11.8,24h.214A12.111,12.111,0,0,0,24,11.791h0A11.766,11.766,0,0,0,11.983,0ZM10.5,16.542a1.476,1.476,0,0,1,1.449-1.53h.027a1.527,1.527,0,0,1,1.523,1.47,1.475,1.475,0,0,1-1.449,1.53h-.027A1.529,1.529,0,0,1,10.5,16.542ZM11,12.5v-6a1,1,0,0,1,2,0v6a1,1,0,1,1-2,0Z"
            fill="currentColor"
          />
        </svg>
      </span>
      <div
        class="chakra-alert__title css-1qkvt8e"
        data-testid="alert-message"
      >
        Error!!!
      </div>
      <button
        aria-label="Close"
        class="css-1mduyzg"
        type="button"
      >
        <svg
          aria-hidden="true"
          class="chakra-icon css-onkibi"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M.439,21.44a1.5,1.5,0,0,0,2.122,2.121L11.823,14.3a.25.25,0,0,1,.354,0l9.262,9.263a1.5,1.5,0,1,0,2.122-2.121L14.3,12.177a.25.25,0,0,1,0-.354l9.263-9.262A1.5,1.5,0,0,0,21.439.44L12.177,9.7a.25.25,0,0,1-.354,0L2.561.44A1.5,1.5,0,0,0,.439,2.561L9.7,11.823a.25.25,0,0,1,0,.354Z"
            fill="currentColor"
          />
        </svg>
      </button>
    </div>
  </div>
  <button
    class="chakra-button css-1twjws8"
    data-testid="sign-in-btn"
    type="submit"
  >
    Sign In
  </button>
</div>
`;

exports[`ForgotPassword should unsuccessfully send the reset email in ForgotPassword 1`] = `
<div
  class="css-1t8zqkj"
>
  <div
    class="css-15sedz4"
  >
    <p
      class="chakra-text css-13gk006"
    >
      Forgot Your Password?
    </p>
    <div
      class="css-ebxf5s"
    >
      <div
        class="chakra-form-control css-0"
        role="group"
      >
        <label
          class="chakra-form__label css-uvho5g"
          for="field-6"
          id="field-6-label"
        >
          Enter your email
        </label>
        <input
          class="chakra-input css-0"
          data-testid="forgot-email"
          id="field-6"
          placeholder="Email"
          type="email"
        />
      </div>
    </div>
    <div
      class="css-g89h0y"
    >
      <div
        class="chakra-alert css-1hwd9xd"
        data-testid="error"
        role="alert"
      >
        <span
          class="chakra-alert__icon css-1c6vscd"
        >
          <svg
            class="chakra-icon css-fi0ww0"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M11.983,0a12.206,12.206,0,0,0-8.51,3.653A11.8,11.8,0,0,0,0,12.207,11.779,11.779,0,0,0,11.8,24h.214A12.111,12.111,0,0,0,24,11.791h0A11.766,11.766,0,0,0,11.983,0ZM10.5,16.542a1.476,1.476,0,0,1,1.449-1.53h.027a1.527,1.527,0,0,1,1.523,1.47,1.475,1.475,0,0,1-1.449,1.53h-.027A1.529,1.529,0,0,1,10.5,16.542ZM11,12.5v-6a1,1,0,0,1,2,0v6a1,1,0,1,1-2,0Z"
              fill="currentColor"
            />
          </svg>
        </span>
        <div
          class="chakra-alert__title css-1qkvt8e"
          data-testid="error-message"
        >
          Error!!!
        </div>
        <button
          aria-label="Close"
          class="css-1mduyzg"
          type="button"
        >
          <svg
            aria-hidden="true"
            class="chakra-icon css-onkibi"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M.439,21.44a1.5,1.5,0,0,0,2.122,2.121L11.823,14.3a.25.25,0,0,1,.354,0l9.262,9.263a1.5,1.5,0,1,0,2.122-2.121L14.3,12.177a.25.25,0,0,1,0-.354l9.263-9.262A1.5,1.5,0,0,0,21.439.44L12.177,9.7a.25.25,0,0,1-.354,0L2.561.44A1.5,1.5,0,0,0,.439,2.561L9.7,11.823a.25.25,0,0,1,0,.354Z"
              fill="currentColor"
            />
          </svg>
        </button>
      </div>
    </div>
    <div
      class="css-1958zak"
    >
      <button
        class="chakra-button css-1fibvam"
        type="submit"
      >
         
        Reset
         
      </button>
    </div>
  </div>
</div>
`;
